### PR TITLE
Add Mutex on Top of p2p Feeds Map

### PR DESCRIPTION
**Pull request by fgimenez**
While working on #316, after enabling the commented proposer tests I'm eventually getting:
```
fatal error: concurrent map writes

goroutine 96 [running]:
runtime.throw(0xe5e088, 0x15)
	GOROOT/src/runtime/panic.go:616 +0x81 fp=0xc4204a9c50 sp=0xc4204a9c30 pc=0x42e1e1
runtime.mapassign(0xcf8760, 0xc420313620, 0xc4204a9d48, 0x171bd00)
	GOROOT/src/runtime/hashmap.go:607 +0x553 fp=0xc4204a9ce0 sp=0xc4204a9c50 pc=0x40b113
github.com/prysmaticlabs/prysm/shared/p2p.(*Server).Feed(0xc42004ccc0, 0xe25860, 0xdd11e0, 0x0)
	shared/p2p/feed.go:41 +0x176 fp=0xc4204a9d78 sp=0xc4204a9ce0 pc=0xb66ef6
github.com/prysmaticlabs/prysm/shared/p2p.(*Server).subscribeToTopic(0xc42004ccc0, 0x1, 0x10659c0, 0xdd11e0)
	shared/p2p/service.go:141 +0x17f fp=0xc4204a9fc0 sp=0xc4204a9d78 pc=0xb6800f
runtime.goexit()
	bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/src/runtime/asm_amd64.s:2361 +0x1 fp=0xc4204a9fc8 sp=0xc4204a9fc0 pc=0x45d491
created by github.com/prysmaticlabs/prysm/shared/p2p.(*Server).Start
	shared/p2p/service.go:82 +0x1dc
```
These changes aim to fix this problem by adding a mutex to prevent concurrent writes to the feeds map. With them in place no more issues of this kind are reported.

### CODE REVIEW

File: WORKSPACE [hunk 1]
```diff
@@ -2,8 +2,8 @@ load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 
 http_archive(
     name = "io_bazel_rules_go",
-    urls = ["https://github.com/bazelbuild/rules_go/releases/download/0.13.0/rules_go-0.13.0.tar.gz"],
```
Comment (nisdas) [inline WORKSPACE]: Why was this changed ?

**Comment by fgimenez** [comment]
@nisdas thanks a lot for your review! :)

After enabling the race detector I was having the same issues as in https://github.com/bazelbuild/rules_go/issues/1592, all goes fine with `0.12.1`. Looks like this will be addressed in https://github.com/bazelbuild/rules_go/pull/1595. Is the codebase relying on any feature specific to `0.13.0`?

Another consequence of enabling the race detector for `//shared/p2p` is that more race conditions are detected other than the concurrent write on the feed map. Not sure if it's ok to handle them on this PR, WDYT?

Also, I'm not sure why the link process is failing on travis, works fine locally. Any ideas?

**Comment by nisdas** [comment]
@fgimenez Ah k , alright got it. I think we might be using features specific to `0.13.0` but I am not too sure. @prestonvanloon What do you think ?
It looks like your tests are failing because its unable to create the container image. 
Yeah unfortunately there are multiple race conditions for #316, it has to do with goroutine leaks due to the multicast DNS server being unable to be shutdown cleanly. 

Maybe it would be better to open up another PR for the other race conditions, do you have any solutions in mind for that ?

**Comment by fgimenez** [comment]
> Yeah unfortunately there are multiple race conditions for #316, it has to do with goroutine leaks due to the multicast DNS server being unable to be shutdown cleanly among other reasons

Yep, I'm seeing those. I've just pushed changes to restrict the race detection to just the test for the feed map concurrent write, PTAL.

> Maybe it would be better to open up another PR for the other race conditions

:100: 

> do you have any solutions in mind for that ?

Not yet! :) But would love to give it a try.

**Comment by nisdas** [comment]
> Not yet! :) But would love to give it a try.

Alright sounds great :+1:  Open up another PR if you find something :)

File: WORKSPACE (hunk 1 above)
Comment (terenc3t) [reply WORKSPACE]: +1, I think this was to fix the mac issue we ran into a month back

File: .travis-bazelrc [hunk 2]
```diff
@@ -10,9 +10,9 @@ build --spawn_strategy=standalone --genrule_strategy=standalone
 #build:remote --google_credentials=/tmp/service-account.json
 
 # Set some build options for travis container.
-build --local_resources=1536,1.5,0.5
```
Comment (terenc3t) [inline .travis-bazelrc]: any reasons on why we need to increase the computing resources here?

Comment (fgimenez) [reply .travis-bazelrc]: nope, I'm trying to debug the build faillures on travis, locally without using `.travis-bazelrc` all goes fine. But even with increased resources it keeps failing https://travis-ci.org/prysmaticlabs/prysm/jobs/408046583. Reverted.

File: WORKSPACE (hunk 1 above)
Comment (fgimenez) [reply WORKSPACE]: @terenc3t about the reasons for this change see https://github.com/prysmaticlabs/prysm/pull/321#issuecomment-407689242, if `0.13.0` is required i'lll remove the concurrent write unit test and the disable the race detector again, it can be added once `0.13.0` (or a higher version) supports it.

File: shared/p2p/feed.go [hunk 3]
```diff
@@ -35,8 +35,10 @@ func (s *Server) Feed(msg interface{}) *event.Feed {
 		t = reflect.TypeOf(msg)
 	}
 
+	s.mutex.Lock()
```
Comment (prestonvanloon) [inline shared/p2p/feed.go]: Please add `defer s.mutex.Unlock()` on the next line and remove the other unlock. This makes it more clear that you’re releasing the locked resource.

File: shared/p2p/BUILD.bazel [hunk 4]
```diff
@@ -1,5 +1,7 @@
 load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
 
+# gazelle:exclude feed_concurrent_test.go
```
Comment (prestonvanloon) [inline shared/p2p/BUILD.bazel]: Please move this next to the target definition and add a comment why this is excluded.

File: WORKSPACE (hunk 1 above)
Comment (prestonvanloon) [reply WORKSPACE]: Please add a comment stating that rules_go should not be updated until that specific bug has been resolved.

The Mac issue that others were referring to had to do with bazel itself, not these rules.

**Comment by prestonvanloon** [comment]
Why did you revert changes in 883ed90c77d1bfeda91950c213553ee8f2f8b27a ?

**Comment by fgimenez** [comment]
> Why did you revert changes in 883ed90 ?

@prestonvanloon Sorry, didn't see your comments and was going ahead with the suggestions for going back to rules_go `0.13.0`, will push shortly the previous status + your comments addressed.

**Comment by fgimenez** [comment]
@prestonvanloon btw, do you know why the build might be failing with `0.12.1` on travis? https://travis-ci.org/prysmaticlabs/prysm/jobs/408046583

File: shared/p2p/feed.go (hunk 3 above)
Comment (fgimenez) [reply shared/p2p/feed.go]: Sure, done.

File: shared/p2p/BUILD.bazel (hunk 4 above)
Comment (fgimenez) [reply shared/p2p/BUILD.bazel]: Done, PTAL.

File: WORKSPACE (hunk 1 above)
Comment (fgimenez) [reply WORKSPACE]: Ok, added comment with a link to the bug.

**Comment by fgimenez** [comment]
yes, `0.12.1` works fine locally

File: shared/p2p/BUILD.bazel [hunk 5]
```diff
@@ -29,6 +29,10 @@ go_library(
     ],
 )
 
+# by default gazelle tries to add all the test files to the  first
+# go_test; we want to treat feed_concurrent_test.go differently
+# gazelle:exclude feed_concurrent_test.go
```
Comment (prestonvanloon) [inline shared/p2p/BUILD.bazel]: Please move this to be next to the go test on line 61, unless gazelle requires it at the top up here for some reason

Comment (fgimenez) [reply shared/p2p/BUILD.bazel]: Sure done, locally `bazel run //:gazelle -- fix` didn't try to change anything.
